man/man1/jpg.1
.TH JPG 1
.SH NAME
jpg, gif, png, ppm, bmp, yuv, ico, togif, toppm, topng, toico \- view and convert pictures
.SH SYNOPSIS
.B jpg
[
.B -39cdefFkJrtv
] [
.I file ...
]
.br
.B gif
[
.B -39cdektv
] [
.I file ...
]
.br
.B png
[
.B -39cdektv
] [
.I file ...
]
.br
.B ppm
[
.B -39cdektv
] [
.I file ...
]
.br
.B bmp
[
.I file
]
.br
.B yuv
[
.I file
]
.PP
.B togif
[
.B -c
.I comment
] [
.B -l
.I loopcount
] [
.B -d
.I msec
] [
.B -t
.I transindex
] [
.I file ...
[
.B -d
.I msec
]
.I file ...
]
.br
.B toppm
[
.B -c
.I comment
] [
.I file
]
.br
.B topng
[
.B -c
.I comment
] [
[
.B -g
.I gamma
] [
.I file
]
.PP
.B ico
[
.I file
]
.br
.B toico
[
.I file ...
]
.SH DESCRIPTION
These programs read, display, and write image files in public formats.
.IR Jpg ,
.IR gif ,
.IR png ,
.IR ppm ,
.IR bmp ,
and
.IR yuv .
read files in the corresponding formats and, by default, display
them in the current window; options cause them instead to convert the images
to Plan 9 image format and write them to standard output.
.IR Togif ,
.IR Toppm ,
and
.I topng
read Plan 9 images files, convert them to GIF, PPM, or PNG, and write them to standard output.
.PP
The default behavior of
.IR jpg ,
.IR gif ,
and
.IR ppm
is to display the
.IR file ,
or standard input if no file is named.
Once a file is displayed, typing a character causes the program to display the next image.
Typing a
.BR q ,
DEL, or control-D exits the program.
For a more user-friendly interface, use
.MR page 1 ,
which invokes these programs to convert the images to standard format,
displays them, and offers scrolling, panning, and menu-driven navigation among the files.
.PP
These programs share many options:
.TP
.B -e
Disable Floyd-Steinberg error diffusion, which is used to improve the appearance
of images on color-mapped displays, typically with 8 bits per pixel.
Primarily useful for debugging; if the display has true RGB color, the image
will be displayed in full glory.
.TP
.B -k
Convert and display the image as a black and white (really grey-scale) image.
.TP
.B -v
Convert the image to an RGBV color-mapped image, even if the
display has true RGB color.
.TP
.B -d
Suppress display of the image; this is set automatically by
any of the following options:
.TP
.B -c
Convert the image to a Plan 9 representation, as defined by
.MR image 7 ,
and write it to standard output.
.TP
.B -9
Like
.BR -c ,
but produce an uncompressed image.
This saves processing time, particularly when the output is
being piped to another program such as
.MR page 1 ,
since it avoids compression and decompression.
.TP
.B -t
Convert the image, if it is in color, to a true color RGB image.
.TP
.B -3
Like
.BR -t ,
but force the image to RGB even if it is originally grey-scale.
.PD
.PP
.I Jpg
has two extra options used to process the output of the LML
video card:
.TP
.B -f
Merge two adjacent images, which represent the two fields of a video picture,
into a single image.
.TP
.B -F
The input is a motion JPEG file, with multiple images representing frames of the movie.  Sets
.BR -f .
.PD
.PP
The
.IR togif
and
.IR toppm
programs go the other way: they convert from Plan 9 images to GIF and PPM,
and have no display capability.
Both accept an option
.B -c
to set the comment field of the resulting file.
If there is only one input picture,
.I togif
converts the image to GIF format.
If there are many
.IR files ,
though, it will assemble them into an animated GIF file.
The options control this process:
.TP
.BI -l loopcount
By default, the animation will loop forever;
.I loopcount
specifies how many times to loop.
A value of zero means loop forever and a negative value means
to stop after playing the sequence once.
.TP
.BI -d msec
By default, the images are displayed as fast as they can be rendered.
This option specifies the time, in milliseconds, to pause while
displaying the next named
.IR file .
.PP
.I Gif
translates files that contain a `transparency' index by attaching
an alpha channel to the converted image.
.PP
.I Ico
displays a Windows icon (.ico) file.  If no file is
specified,
.I ico
reads from standard input.
Icon files
contain sets of icons represeted by an image and a mask.
Clicking the right button pops up a menu that lets you
write any icon's image as a Plan 9 image (\fIwidth\fBx\fIheight\fB.image),
write any icon's mask as a Plan 9 image (\fIwidth\fBx\fIheight\fB.mask),
or exit.  Selecting one of the write menu items yields a sight cursor.
Move the sight over the icon and right click again to write.
.PP
.I Toico
takes a list of Plan 9 image files (or standard input) and creates
a single icon file.  The masks in the icon file will be the white
space in the image.  The icon file is written to standard output.
.SH SOURCE
.B \*9/src/cmd/jpg
.SH "SEE ALSO"
.MR page 1 ,
.MR image 7 .
.SH BUGS
Writing an animated GIF using
.I togif
is a clumsy undertaking.


CHANGES
# Note: the name listed next to a change sometimes
# submitted the patch but more often just reported
# the problem.

# For entries more recent than January 30, 2008,
# see the Git repository history at 
# https://github.com/9fans/plan9port/.

January 30, 2008
	acme: fix regexp match choice bug
	devdraw: add F11 for full-screen toggle
	rio: accomodate full-screen apps more gracefully
	sam: fix regexp match choice bug

January 28, 2008
	xd: fix printing of 4-byte values on 64-bit machines (Sam Hopkins)

January 12, 2008
	delatex: reset lexer state at each beginning of file
	ipso: do not flush factotum keys unless using factotum file
	libdiskfs: fix ext2 directory reading bug (Mike Mammarella)
	lp: move devices to devices.sample
	rio: make full-screen work properly; add showevent
	venti: fewer calls to gettimeofday
	vmount: need to say -t nfs now on Linux

January 10, 2008
	acme: make STACK bigger for OS X (David Swasey)
	dist,unix: use MKSHELL=$PLAN9/bin/rc (John DeGood)
	libdraw: fix spinning event loop (David Bulkow)
	libregexp: fix match choice bug

December 7, 2007
	acme, sam, libregexp: regexp fixes and reversions

December 4, 2007
	hoc: print exact floating point output
	sam: turn off DEBUG logging (oops)
	vbackup: fix uninitialized variable use in -i code

November 28, 2007
	acme: refresh window properly after Tab command (Fazlul Shahriar)

November 27, 2007
	.hgignore: ignore more
	libdiskfs: make ext2 use little-endian disk format on big-endian machines (Joshua Wood)
	acme: accept clicks between body and scroll bar (Axel Belinfante)
	tpic: clean up better
	upas/smtp: remove intermediate files
	vbackup: add -i flag to avoid score prefetching (David Swasey)
	venti: add debugging HTML form

November 25, 2007
	acme: avoid infinite loop on search in 8125-byte file (Roger Peppe)
	lib9: avoid DIOCGMEDIASIZE on FreeBSD 4.11 (Rodrigo Readi)
	mailfs: use root in imapcopylist (Richard Bilson)
	moveplan9.files: add tmac/tmac.an (Tony Lainson)

November 23, 2007
	9pfuse: ignore Linux O_CLOEXEC flag (Michael Kaminsky)

November 22, 2007
	9pfuse: fix array bounds error (Michael Teichgräber)
	vbackup: fix fsysopen error message (Joshua Wood)

November 7, 2007
	9fs: use -n to mount sources (Jeff Sickel),
	9pfuse: MacFUSE moved again! (Jeff Sickel)

November 6, 2007
	.cvsignore: add 9pm (for old trees)
	unvf: ignore case in Content-Transfer-Encoding: line.
	validateattachment: reject malformed zip files

November 5, 2007
	building: use ar s flag to avoid ranlib (OS X)
	diff: continue after i/o errors when diffing many files
	factotum: silence gcc warning
	getns: turn / into _ in $DISPLAY
	libdiskfs: OS X linker fixes
	libmach: OS X linker fixes
	libmp: no text jump labels (OS X)
	libthread: use our own swapcontext on OS X
	mailfs: handle NIL body, add -r flag (Richard Bilson)

November 4, 2007
	man: document hg interface, fix CVS host name

November 3, 2007
	libthread: more OS X 10.5 changes (Jeff Sickel)

October 31, 2007
	venti: use 9ar, not ar (fixes Darwin build)

October 30, 2007
	libthread: OS X 10.5 tweaks (Jeff Sickel, Bakul Shah)

October 29, 2007
	grap: fclose, error context bug fixes (Doug Gwyn)
	venti: fix sync deadlock, add /proc stub

October 12, 2007
	9p: add -n to disable authentication
	9pserve: add -n to reject authentication
	srv: add -n to reject authentication

October 9, 2007
	mount: ask for Unix test explicitly (Chris King)

September 24-25, 2007
	venti: many fixes, new index cache

September 16, 2007
	factotum: fix build error (Axel Belinfante)

September 15, 2007
	auxstats: use acpi for battery on Linux (Andrey Mirtchovski)

September 13, 2007
	factotum: implement 9pcr (of dubious utility)

August 28, 2007
	9pserve: fix infinite loop bug

August 23, 2007
	venti: add -s flag, bug fixes (Michael Kaminsky)

August 22, 2007
	9.rc: change to work with rc -e (Tony Lainson)
	9pserve: fix races causing assert(c->nmsg == 0) to fail.
	acme: fix warning memory leak (Michael Teichgräber)
	libfmt: add fmtlocale to Makefile (Faried Nawaz)
	libthread: simplify alt tracking, possibly fixing bug
	mailfs: fix handling of FLAGS responses (Robert Vollmert)
	unix ports: Linux x86-64 Makefiles (Ævar Arnfjörð Bjarmason)

July 25, 2007
	9pfuse: even more MacFUSE path changes

June 29, 2007
	dict: look in $dictpath for dictionaries

June 24, 2007
	mailfs: fix subject line in header file.

June 23, 2007
	mailfs: various IMAP bug fixes
	Mail: add Get command, refresh at startup,
		fix deleted messages bug
	nedmail: refresh at startup, on i and y.

June 20, 2007
	9pfuse: add -a aname to get at sourcesdump

June 19, 2007
	9pfuse: the MacFUSE file locations changed again (Rob Pike)

June 18, 2007
	upas/marshal: remove debugging print
	9pserve: ignore window resizes (fix really should be elsewhere)
	libdiskfs: more HFS fixes (David Swasey)

June 15, 2007
	acme(1): fix Snarf bug, change Indent ON to apply to all windows.

June 12, 2007
	vcat: always write last block, to make disk correct size
	vbackup: more OS X fixes, add -M flag
	vbackup(8): document -M

June 9, 2007
	libregexp, sam, acme: day 1 bug in sam regexp code

June 8, 2007
	libdiskfs: add HFS resource fork support (David Swasey)

June 5, 2007
	ramfs: be less chatty

June 3, 2007
	vbackup: fix dotdot (David Swasey)

May 29, 2007
	libdiskfs: more HFS fixes (David Swasey)
	
May 28, 2007
	libdiskfs: missing Bterm, make ext2 quieter,
		add disk partitions, HFS code (David Swasey)

May 21, 2007
	libregexp: fix pointer comparison warning (Ramon de Vera)
	test: fix
	
May 20, 2007
	touch: print error if wstat fails
	libc: use futimes on Linux to implement dirfwstat

May 19, 2007
	test: import from Plan 9

May 18, 2007
	libregexp: fix bug about leading ^ in regexp on multiline text (Roger Peppe)

May 15, 2007
	libc: new media size code for _p9dir on FreeBSD (Lou Kamenov)
	acme(4): document tab width field in ctl file

May 10, 2007
	libc: more NaN fixes (rsc)
		use device names in _p9dir for FreeBSD (Lou Kamenov)

src/cmd/jpg/jpg.c
#include <u.h>
#include <libc.h>
#include <bio.h>
#include <draw.h>
#include <event.h>
#include "imagefile.h"

int		cflag = 0;
int		dflag = 0;
int		eflag = 0;
int		jflag = 0;
int		fflag = 0;
int		Fflag = 0;
int		nineflag = 0;
int		threeflag = 0;
int		colorspace = CYCbCr;	/* default for 8-bit displays: combine color rotation with dither */
int		output = 0;
ulong	outchan = CMAP8;
Image	*image;
int		defaultcolor = 1;

enum{
	Border	= 2,
	Edge		= 5
};

char	*show(int, char*, int);

void
eresized(int new)
{
	Rectangle r;

	if(new && getwindow(display, Refnone) < 0){
		fprint(2, "jpg: can't reattach to window\n");
		exits("resize");
	}
	if(image == nil)
		return;
	r = rectaddpt(image->clipr, subpt(screen->r.min, image->clipr.min));
	if(!new && !winsize)
		drawresizewindow(r);
	draw(screen, r, image, nil, image->r.min);
	flushimage(display, 1);
}

void
usage(void)
{
	fprint(2, "usage: jpg -39cdefFkJrtv -W winsize [file.jpg ...]\n");
	exits("usage");
}

void
main(int argc, char *argv[])
{
	int fd, i, yflag;
	char *err;
	char buf[12+1];

	yflag = 0;
	ARGBEGIN{
	case 'W':
		winsize = EARGF(usage());
		break;
	case 'c':		/* produce encoded, compressed, bitmap file; no display by default */
		cflag++;
		dflag++;
		output++;
		if(defaultcolor)
			outchan = CMAP8;
		break;
	case 'd':		/* suppress display of image */
		dflag++;
		break;
	case 'e':		/* disable floyd-steinberg error diffusion */
		eflag++;
		break;
	case 'F':
		Fflag++;	/* make a movie */
		fflag++;	/* merge two fields per image */
		break;
	case 'f':
		fflag++;	/* merge two fields per image */
		break;
	case 'J':		/* decode jpeg only; no display or remap (for debugging, etc.) */
		jflag++;
		break;
	case 'k':		/* force black and white */
		defaultcolor = 0;
		outchan = GREY8;
		break;
	case 'r':
		colorspace = CRGB;
		break;
	case '3':		/* produce encoded, compressed, three-color bitmap file; no display by default */
		threeflag++;
		/* fall through */
	case 't':		/* produce encoded, compressed, true-color bitmap file; no display by default */
		cflag++;
		dflag++;
		output++;
		defaultcolor = 0;
		outchan = RGB24;
		break;
	case 'v':		/* force RGBV */
		defaultcolor = 0;
		outchan = CMAP8;
		break;
	case 'y':	/* leave it in CYCbCr; for debugging only */
		yflag = 1;
		colorspace = CYCbCr;
		break;
	case '9':		/* produce plan 9, uncompressed, bitmap file; no display by default */
		nineflag++;
		dflag++;
		output++;
		if(defaultcolor)
			outchan = CMAP8;
		break;
	default:
		usage();
	}ARGEND;

	if(yflag==0 && dflag==0 && colorspace==CYCbCr){	/* see if we should convert right to RGB */
		fd = open("/dev/screen", OREAD);
		if(fd > 0){
			buf[12] = '\0';
			if(read(fd, buf, 12)==12 && chantodepth(strtochan(buf))>8)
				colorspace = CRGB;
			close(fd);
		}
	}

	err = nil;
	if(argc == 0)
		err = show(0, "<stdin>", outchan);
	else{
		for(i=0; i<argc; i++){
			fd = open(argv[i], OREAD);
			if(fd < 0){
				fprint(2, "jpg: can't open %s: %r\n", argv[i]);
				err = "open";
			}else{
				err = show(fd, argv[i], outchan);
				close(fd);
			}
			if((nineflag || cflag) && argc>1 && err==nil){
				fprint(2, "jpg: exiting after one file\n");
				break;
			}
		}
	}
	exits(err);
}

Rawimage**
vidmerge(Rawimage **aa1, Rawimage **aa2)
{
	Rawimage **aao, *ao, *a1, *a2;
	int i, c, row, col;

	aao = nil;
	for (i = 0; aa1[i]; i++) {

		a1 = aa1[i];
		a2 = aa2[i];
		if (a2 == nil){
			fprint(2, "jpg: vidmerge: unequal lengths\n");
			return nil;
		}
		aao = realloc(aao, (i+2)*sizeof(Rawimage *));
		if (aao == nil){
			fprint(2, "jpg: vidmerge: realloc\n");
			return nil;
		}
		aao[i+1] = nil;
		ao = aao[i] = malloc(sizeof(Rawimage));
		if (ao == nil){
			fprint(2, "jpg: vidmerge: realloc\n");
			return nil;
		}
		memcpy(ao, a1, sizeof(Rawimage));
		if (!eqrect(a1->r , a2->r)){
			fprint(2, "jpg: vidmerge: rects different in img %d\n", i);
			return nil;
		}
		if (a1->cmaplen != a2->cmaplen){
			fprint(2, "jpg: vidmerge: cmaplen different in img %d\n", i);
			return nil;
		}
		if (a1->nchans != a2->nchans){
			fprint(2, "jpg: vidmerge: nchans different in img %d\n", i);
			return nil;
		}
		if (a1->fields != a2->fields){
			fprint(2, "jpg: vidmerge: fields different in img %d\n", i);
			return nil;
		}
		ao->r.max.y += Dy(ao->r);
		ao->chanlen += ao->chanlen;
		if (ao->chanlen != Dx(ao->r)*Dy(ao->r)){
			fprint(2, "jpg: vidmerge: chanlen wrong %d != %d*%d\n",
				ao->chanlen, Dx(ao->r), Dy(ao->r));
			return nil;
		}
		row = Dx(a1->r);
		for (c = 0; c < ao->nchans; c++) {
			uchar *po, *p1, *p2;

			ao->chans[c] = malloc(ao->chanlen);
			po = ao->chans[c];
			p1 = a1->chans[c];
			p2 = a2->chans[c];
			for (col = 0; col < Dy(a1->r); col++) {
				memcpy(po, p1, row);
				po += row, p1 += row;
				memcpy(po, p2, row);
				po += row, p2 += row;
			}
			free(a1->chans[c]);
			free(a2->chans[c]);
		}
		if(a2->cmap != nil)
			free(a2->cmap);
		free(a1);
		free(a2);
	}
	if (aa2[i] != nil)
		fprint(2, "jpg: vidmerge: unequal lengths\n");
	free(aa1);
	free(aa2);
	return aao;
}

char*
show(int fd, char *name, int outc)
{
	Rawimage **array, *r, *c;
	static int inited;
	Image *i;
	int j, ch, outchan;
	Biobuf b;
	char buf[32];

	if(Binit(&b, fd, OREAD) < 0)
		return nil;
	outchan = outc;
rpt:	array = Breadjpg(&b, colorspace);
	if(array == nil || array[0]==nil){
		fprint(2, "jpg: decode %s failed: %r\n", name);
		return "decode";
	}
	if (fflag) {
		Rawimage **a;

		a = Breadjpg(&b, colorspace);
		if(a == nil || a[0]==nil){
			fprint(2, "jpg: decode %s-2 failed: %r\n", name);
			return "decode";
		}
		array = vidmerge(a, array);
	} else
		Bterm(&b);

	r = array[0];
	c = nil;
	if(jflag)
		goto Return;
	if(!dflag){
		if(!inited){
			if(initdraw(0, 0, 0) < 0){
				fprint(2, "jpg: initdraw failed: %r\n");
				return "initdraw";
			}
			if(Fflag == 0)
				einit(Ekeyboard|Emouse);
		inited++;
		}
		if(defaultcolor && screen->depth>8 && outchan==CMAP8)
			outchan = RGB24;
	}
	if(outchan == CMAP8)
		c = torgbv(r, !eflag);
	else{
		if(outchan==GREY8 || (r->chandesc==CY && threeflag==0)){
			c = totruecolor(r, CY);
			outchan = GREY8;
		}else
			c = totruecolor(r, CRGB24);
	}
	if(c == nil){
		fprint(2, "jpg: conversion of %s failed: %r\n", name);
		return "torgbv";
	}
	if(!dflag){
		if(c->chandesc == CY)
			i = allocimage(display, c->r, GREY8, 0, 0);
		else
			i = allocimage(display, c->r, outchan, 0, 0);
		if(i == nil){
			fprint(2, "jpg: allocimage %s failed: %r\n", name);
			return "allocimage";
		}
		if(loadimage(i, i->r, c->chans[0], c->chanlen) < 0){
			fprint(2, "jpg: loadimage %s failed: %r\n", name);
			return "loadimage";
		}
		image = i;
		eresized(0);
		if (Fflag) {
			freeimage(i);
			for(j=0; j<r->nchans; j++)
				free(r->chans[j]);
			free(r->cmap);
			free(r);
			free(array);
			goto rpt;
		}
		if((ch=ekbd())=='q' || ch==0x7F || ch==0x04)
			exits(nil);
		draw(screen, screen->clipr, display->white, nil, ZP);
		image = nil;
		freeimage(i);
	}
	if(nineflag){
		chantostr(buf, outchan);
		print("%11s %11d %11d %11d %11d ", buf,
			c->r.min.x, c->r.min.y, c->r.max.x, c->r.max.y);
		if(write(1, c->chans[0], c->chanlen) != c->chanlen){
			fprint(2, "jpg: %s: write error %r\n", name);
			return "write";
		}
	}else if(cflag){
		if(writerawimage(1, c) < 0){
			fprint(2, "jpg: %s: write error: %r\n", name);
			return "write";
		}
	}
    Return:
	for(j=0; j<r->nchans; j++)
		free(r->chans[j]);
	free(r->cmap);
	free(r);
	free(array);
	if(c){
		free(c->chans[0]);
		free(c);
	}
	if (Fflag) goto rpt;
	return nil;
}
